Intermediate macro blocks

nibhar opened the pull request:
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.

## What's in this pull request?

This pull request adds intermediate macro blocks as the third block type. They work the same as macro blocks do, with the exception that the validator set does not change.

nibhar committed:
Add batch to policy, add election functions to policy

nibhar committed:
add election block notion to macro blocks, adapt to batches and election

blocks
also: cargo fmt

paberr commented:
While refactoring the whole slash registry/reward registry thing, I realised that we *do* need to include both slashed sets in the election blocks.

Previous to this PR, election blocks only contained the slashed set of the previous epoch (not the current one that ends with the block at hand). This was and is required for syncing, so that rewards can be calculated correctly.

This PR now added the current slashed set to macro blocks that are no election blocks. However, consider the following scenario: I synced up to the latest election block and start producing blocks. The election block didn't include the "current epoch" slashed set (which now after beginning the next epoch is the "previous epoch" slashed set). So, upon arriving at the first checkpoint block, this information is missing.

I'll address this issue in my follow up PR that refactors said code.

brunoffranca committed:
1. Added documentation.
2. Fixed error regarding extra data field in the macro block production.
3. Simplified block-production-albatross/lib.rs.

brunoffranca commented on blockchain/src/blockchain/mod.rs:
```diff
@@ -1004,12 +1002,20 @@ impl AbstractBlockchain for Blockchain {
             .get_chunk(prefix, size, txn_option)
     }
 
+    fn get_batch_transactions(
+        &self,
+        batch: u32,
+        txn_option: Option<&Transaction>,
+    ) -> Option<Vec<BlockchainTransaction>> {
+        self.get_epoch_transactions(batch, txn_option)
+    }
+
```
What does this actually do? It seems like it returns the transactions for epoch x when you ask for the transactions for batch x.

nibhar replied:
This is part of the 1.0 chain object. Therefore there is no concept of epochs or batches and the functions threat the input as the actual block height, effectively making both functions exactly the same.

It is a remnant of the single repo for both versions of the protocol.

brunoffranca replied:
Ok. Seems good to me then.

brunoffranca commented:
We only need to solve the issue of failing when we provide any extra data to the macro blocks and then it can be merged.

nibhar committed:
Correctly add MacroBlock extriniscs.extra_data before verifying its hash

nibhar committed:
cargo fmt
